File: 백준/Silver/1874. 스택 수열/스택 수열.js
const fs = require('fs');
const filePath =
  process.platform === 'linux' ? '/dev/stdin' : '../../input.txt';
const input = fs.readFileSync(filePath).toString().trim().split('\n');
const N = Number(input[0]);
const info = input.slice(1).map(Number);
const stack = [];
const result = [];
let current = 1;
let possible = true;

for(let i = 0; i < info.length; i++) {
    const target = info[i];
    while(current <= target) {
        stack.push(current);
        result.push('+');
        current++;
    }

    const popped = stack.pop();
    if(popped !== target) {
        possible = false;
        break;
    }
    result.push('-');
}

console.log(possible ? result.join('\n') : 'NO')

File: 백준/Silver/1874. 스택 수열/스택 수열.py
import sys
from collections import deque
input = sys.stdin.read().splitlines()
n = int(input[0])
nums = deque(list(map(int, input[1:])))

def solve(n, nums):
    stack = []
    result = []
    num = 1
    while nums:
        target = nums[0]
    
        if not stack or stack[-1] < target:
            stack.append(num)
            result.append('+')
            if num < n:
                num += 1
        elif stack[-1] == target:
            stack.pop()
            result.append('-')
            nums.popleft()
        elif stack[-1] > target:
            return "NO"
    return '\n'.join(result)
    
print(solve(n, nums))
